### PR TITLE
EZP-30931: Skipped archiving URL Alias when moving identical Locations

### PR DESCRIPTION
**Pull request by ITernovtsii**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30931](https://jira.ez.no/browse/EZP-30931)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | 7.5
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

Child location broken after location moved under the same parent.
It's caused by `eZ\Publish\Core\Persistence\Legacy\Content\UrlAlias\Handler::historizeId` function executed after same alias reused.
This PR prevents location to be moved to the same parent (from the core or user code).

Additionally, I added checking of `isContainer` status to be performed in the same place. So, it will be possible to remove it from https://github.com/ezsystems/ezplatform-admin-ui/blob/v1.5.6/src/bundle/Controller/LocationController.php#L130-L135
Most test changes are related to moving under a non-container location (Demo Design).

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php [hunk 1]
```diff
@@ -252,6 +252,9 @@ public function locationMoved($locationId, $oldParentId, $newParentId)
                 'newParent' => $newParentId,
             ]
         );
+        if ($oldParentId === $newParentId) {
+            return $newParentId;
+        }
```
Comment (andrerom) [inline eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: Don't think this logic belongs in cache, only side effect here is that cache gets cleared anyway.

Comment (ITernovtsii) [reply eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: agree, is it okay if I'll remove this part here and do this instead?
(few lines below):
```
if ($oldParentId !== $newParentId) {
    $this->cache->invalidateTags(['urlAlias-location-' . $locationId, 'urlAlias-location-path-' . $locationId]);
}
```
Or, may I just remove it? It will not bring any issues, as it's not something you would do often)

@andrerom  what would you prefer?

Comment (andrerom) [reply eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: what you suggest is ok, the risk is if the `$oldParentId !== $newParentId` logic changes but if so this code should be adapted too.

Comment (ITernovtsii) [reply eZ/Publish/Core/Persistence/Cache/UrlAliasHandler.php]: done

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 2]
```diff
@@ -2348,19 +2348,19 @@ public function testMoveSubtree()
         $locationService = $repository->getLocationService();
 
         // Load location to move
-        $locationToMove = $locationService->loadLocation($mediaLocationId);
+        $locationToMove = $locationService->loadLocation($demoDesignLocationId);
```
Comment (andrerom) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: Was these changes there from before? Could we drop them (changes in `testMoveSubtree()`)? (seems unrelated)

Comment (andrerom) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: same goes for many of the other test changes below

Comment (ITernovtsii) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: Hi @andrerom 
Additionally, I added checking of isContainer status to be performed in the same place.
So, it will be possible to remove it from https://github.com/ezsystems/ezplatform-admin-ui/blob/v1.5.6/src/bundle/Controller/LocationController.php#L130-L135
Most test changes are related to moving under a non-container location (Demo Design).

Not sure why that validation was placed in the admin-ui bundle initially.
If there was a reason for that, I can remove `isContainer` part here.
But I think it still makes sense to have tests updated, so the content will be moved under node with isContainer=true.

Comment (andrerom) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: The addition of `isContainer ` check is good, on that just make sure to add PHP doc on the interface to describe the new exception case.

My comment was purely about the test changes in `testMoveSubtree()` and the existing `testMoveSubtree*()` methods below, are these changes needed to make them pass? Did they move to non contianer?

Comment (ITernovtsii) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: Yes, existing testMoveSubtree is testing moving under "Demo Design", which is not a container (as "success" use-case),. So, I updated them to test moving under "Media" location.

Comment (andrerom) [reply eZ/Publish/API/Repository/Tests/LocationServiceTest.php:2546]: ok

**Review by andrerom — APPROVED** [review]
+1, but we need addition to API php-doc for the new exception case.

**Comment by ITernovtsiy** [comment]
@andrerom please check phpdoc updates https://github.com/ezsystems/ezpublish-kernel/pull/2761/commits/56c7e91eaddd087ad737779532057abddd64bdb9
Let me know if I should replace inheritDoc to actual comment.

File: eZ/Publish/Core/Repository/LocationService.php [hunk 3]
```diff
@@ -650,23 +650,31 @@ public function unhideLocation(APILocation $location)
...
-     * If a user has the permission to move the location to a target location
-     * he can do it regardless of an existing descendant on which the user has no permission.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to move this location to the target
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user does not have read access to the whole source subtree
-     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException If the new parent is in a subtree of the location
-     *
-     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
-     * @param \eZ\Publish\API\Repository\Values\Content\Location $newParentLocation
+     * @inheritDoc
```
Comment (andrerom) [inline eZ/Publish/Core/Repository/LocationService.php]:
```suggestion
     * {@inheritdoc}
```

File: eZ/Publish/Core/REST/Client/LocationService.php [hunk 4]
```diff
@@ -357,15 +357,7 @@ public function copySubtree(Location $subtree, Location $targetParentLocation)
...
-     * Moves the subtree to $newParentLocation.
-     *
-     * If a user has the permission to move the location to a target location
-     * he can do it regardless of an existing descendant on which the user has no permission.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user user is not allowed to move this location to the target
-     *
-     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
-     * @param \eZ\Publish\API\Repository\Values\Content\Location $newParentLocation
+     * @inheritDoc
```
Comment (andrerom) [inline eZ/Publish/Core/REST/Client/LocationService.php]:
```suggestion
     * {@inheritdoc}
```

**Review by andrerom — APPROVED** [review]

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 5]
```diff
@@ -2597,17 +2598,78 @@ public function testMoveSubtreeHidden()
...
         $newParentLocation = $locationService->loadLocation($demoDesignLocationId);
 
+        // Move location from "Home" to "Demo Design" (not container)
+        $locationService->moveSubtree($locationToMove, $newParentLocation);
+    }
+
+    /**
+     * Test for the moveSubtree() method.
+     *
+     * @see \eZ\Publish\API\Repository\LocationService::moveSubtree()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: Please use `@covers` instead for all new or changed test cases, without method call parentheses, like that:
```suggestion
     * @covers \eZ\Publish\API\Repository\LocationService::moveSubtree
```

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 6]
```diff
@@ -2597,17 +2598,78 @@ public function testMoveSubtreeHidden()
...
+     * @depends eZ\Publish\API\Repository\Tests\LocationServiceTest::testLoadLocation
+     * @expectedException \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: This is deprecated and won't work on `ezplatform-kernel:1.x`. Use `$this->expectException` instead please, but be careful when placing it - it should be located just before the call which is supposed to throw that exception.

File: eZ/Publish/API/Repository/Tests/LocationServiceTest.php [hunk 7]
```diff
@@ -2597,17 +2598,78 @@ public function testMoveSubtreeHidden()
...
+     */
+    public function testMoveSubtreeThrowsExceptionOnMoveToSame()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/LocationServiceTest.php]: At least for all new test cases (methods), ideally for all the affected since this should not pollute the diff too much.
```suggestion
    public function testMoveSubtreeThrowsExceptionOnMoveToSame(): void
```

File: eZ/Publish/Core/Repository/LocationService.php [hunk 8]
```diff
@@ -654,23 +654,31 @@ public function unhideLocation(APILocation $location)
...
+            throw new InvalidArgumentException(
+                '$newParentLocation',
+                'new parent location is in a subtree of the given $location'
+            );
+        }
+        if (!$newParentLocation->getContent()->getContentType()->isContainer) {
+            throw new InvalidArgumentException(
+                '$newParentLocation',
+                'Cannot move location to a parent that is not a container'
+            );
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/LocationService.php]: I need someone from @ezsystems/documentation-team too look at those messages.

**Review by alongosz — CHANGES_REQUESTED** [review]
@ITernovtsiy I've rebased PR to make sure it still works.

What I don't get is the "Demo design" / "Home" / "Media" changes to tests. What is the reason for that?

Other remarks:

**Comment by ITernovtsii** [comment]
@alongosz Done, updated `see` to `covers` , `@expectedException` to `$this->expectException`, and appended `: void`.

> What I don't get is the "Demo design" / "Home" / "Media" changes to tests. What is the reason for that?

Those tests are failing after adding "isContainer" check, as "Demo Design" location has `isContainer = false` _(and current tests are moving other location under it)_.
So, I changed them to use the proper location, which is a container.

**Review by alongosz — APPROVED** [review]

File: eZ/Publish/Core/Repository/LocationService.php [hunk 9]
```diff
@@ -654,23 +654,31 @@ public function unhideLocation(APILocation $location)
...
+     * {@inheritdoc}
      */
     public function moveSubtree(APILocation $location, APILocation $newParentLocation)
     {
         $location = $this->loadLocation($location->id);
         $newParentLocation = $this->loadLocation($newParentLocation->id);
 
+        if ($newParentLocation->id === $location->parentLocationId) {
+            throw new InvalidArgumentException(
+                '$newParentLocation', 'new parent location is the same as current'
```
Comment (DominikaK) [inline eZ/Publish/Core/Repository/LocationService.php]:
```suggestion
                '$newParentLocation', 'new parent Location is the same as the current one'
```

File: eZ/Publish/Core/Repository/LocationService.php [hunk 10]
```diff
@@ -654,23 +654,31 @@ public function unhideLocation(APILocation $location)
...
+            );
+        }
+        if (strpos($newParentLocation->pathString, $location->pathString) === 0) {
+            throw new InvalidArgumentException(
+                '$newParentLocation',
+                'new parent location is in a subtree of the given $location'
```
Comment (DominikaK) [inline eZ/Publish/Core/Repository/LocationService.php]:
```suggestion
                'new parent Location is in a subtree of the given $location'
```

File: eZ/Publish/Core/Repository/LocationService.php [hunk 11]
```diff
@@ -654,23 +654,31 @@ public function unhideLocation(APILocation $location)
...
+            );
+        }
+        if (!$newParentLocation->getContent()->getContentType()->isContainer) {
+            throw new InvalidArgumentException(
+                '$newParentLocation',
+                'Cannot move location to a parent that is not a container'
```
Comment (DominikaK) [inline eZ/Publish/Core/Repository/LocationService.php]:
```suggestion
                'Cannot move Location to a parent that is not a container'
```

**Review by tomaszszopinski — APPROVED** [review]
QA approved on eZPlatform-ee 2.5.12 & 2.5.13 with diff & patch.

**Comment by alongosz** [comment]
Thank you @ITernovtsiy
